File: swift/Sources/MossFormer2SR/Snake.swift
import MLX
import MLXNN

/// Snake activation function
public func snake(_ x: MLXArray, alpha: MLXArray) -> MLXArray {
    return x + MLX.reciprocal(alpha + 1e-9) * MLX.pow(MLX.sin(alpha * x), 2)
}

/// Snake1d activation module
public class Snake1d: Module {
    @ModuleInfo var alpha: MLXArray
    
    public init(channels: Int) {
        self.alpha = MLXArray.ones([1, channels, 1])
        super.init()
    }
    
    public func callAsFunction(_ x: MLXArray) -> MLXArray {
        return snake(x, alpha: alpha)
    }
}

File: python/snake.py
import mlx.core as mx
import mlx.nn as nn

def snake(x, alpha):
    """Snake activation function"""
    return x + mx.reciprocal(alpha + 1e-9) * mx.power(mx.sin(alpha * x), 2)

class Snake1d(nn.Module):
    def __init__(self, channels):
        super().__init__()
        self.alpha = mx.ones((1, channels, 1))
    
    def __call__(self, x):
        return snake(x, self.alpha)